lib/value/src/kind/collection.rs

```diff
@@ -0,0 +1,403 @@
+    }
+
+    /// Merge the `other` collection into `self`.
+    ///
+    /// The following merge strategies are applied.
+    ///
+    /// For *known fields*:
+    ///
+    /// - If a field exists in both collections, their `Kind`s are merged, or the `other` fields
+    ///   are used (depending on the configured [`Strategy`](merge::Strategy).
+    ///
+    /// - If a field exists in one but not the other, the field is used.
+    ///
+    /// For *unknown fields or indices*:
+    ///
+    /// - Both `Unknown`s are merged, similar to merging two `Kind`s.
+    pub fn merge(&mut self, mut other: Self, strategy: merge::Strategy) {
+        self.known
+            .iter_mut()
+            .for_each(|(key, self_kind)| match other.known.remove(&key) {
+                Some(other_kind) if strategy.depth.is_shallow() => *self_kind = other_kind,
+                Some(other_kind) => self_kind.merge(other_kind, strategy),
+                _ => {}
+            });
+
+        self.known.extend(other.known.into_iter());
+        self.unknown = self.unknown | other.unknown;
+    }
+}
+
+impl<T: Ord> From<BTreeMap<T, Kind>> for Collection<T> {
+    fn from(known: BTreeMap<T, Kind>) -> Self {
+        Self {
+            known,
+            unknown: Unknown::any(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use super::*;
+
+    #[test]
+    #[allow(clippy::too_many_lines)]
+    fn test_is_superset() {
+        struct TestCase {
+            this: Collection<&'static str>,
+            other: Collection<&'static str>,
+            want: bool,
+        }
+
+        for (title, TestCase { this, other, want }) in HashMap::from([
+            (
+                "any comparison",
+                TestCase {
+                    this: Collection::any(),
+                    other: Collection::any(),
+                    want: true,
+                },
+            ),
+            (
+                "exact/any mismatch",
+                TestCase {
+                    this: Collection::json(),
+                    other: Collection::any(),
+                    want: false,
+                },
+            ),
+            (
+                "unknown match",
+                TestCase {
+                    this: Collection::unknown(Kind::regex().or_null()),
+                    other: Collection::unknown(Kind::regex()),
+                    want: true,
+                },
+            ),
+            (
+                "unknown mis-match",
+                TestCase {
+                    this: Collection::unknown(Kind::regex().or_null()),
+                    other: Collection::unknown(Kind::bytes()),
+                    want: false,
+                },
+            ),
+            (
+                "other-known match",
+                TestCase {
+                    this: Collection::from_parts(
+                        BTreeMap::from([("bar", Kind::bytes())]),
+                        Kind::regex().or_null(),
```

Comment:
Often, the starting states are mostly there for me to mix-and-match different type states so that there aren't any weird bugs that creep in that happen only for one state, but not the other.

Small chance, and usually this would be better tested with property tests and the like, but it's a small price to pay, for a tiny bit of extra security.